lib/Scheduling/ASAPScheduler.cpp

```diff
@@ -0,0 +1,63 @@
+//===- ASAPScheduler.cpp - As-soon-as-possible list scheduler -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of an as-soon-as-possible list scheduler for acyclic problems.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/ASAPScheduler.h"
+
+#include <deque>
+
+using namespace circt;
+using namespace circt::scheduling;
+
+LogicalResult ASAPScheduler::schedule() {
+  if (failed(check()))
+    return failure();
+
+  // initialize a worklist with the block's operations
+  std::deque<Operation *> worklist;
```

Comment:
Could you use a `SmallVector` for this: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h? I'm not sure the overheads of a `deque` are warrented (see that document's recommendations on `deque`). If not a `SmallVector`, perhaps just a `std::vector`? That's how one of the core upstream pattern drivers stores its own worklist, for example: https://github.com/llvm/llvm-project/blob/d4ed965b2d14361e4273525a9888344a25f3800c/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp#L130

Reply:
Yes, I switched to a `SmallVector`, as I no longer needed the queue behavior after implementing the fail-safe.